### PR TITLE
Support Ollama backend.

### PR DESCRIPTION
**Pull request by Isotr0py**
Related issue:
- #49

Features:
- Support `Ollama` backend for server.
- Support auto pulling model from ollama's model repo like [onekuma/sakura-13b-lnovel-v0.9b-q2_k](https://ollama.com/onekuma/sakura-13b-lnovel-v0.9b-q2_k)

TODO:
- [x] Test ollama auto pulling from [onekuma/sakura-13b-lnovel-v0.9b-q2_k](https://ollama.com/onekuma/sakura-13b-lnovel-v0.9b-q2_k).
- [x] Test model loading and inference.

~~Still have some bugs, marked as draft.~~

Deploy (ollama need to be installed before running):
```
python server.py \
    --model_name_or_path onekuma/sakura-13b-lnovel-v0.9b-q2_k \
    --ollama \
    --model_version 0.9 \
    --trust_remote_code \
    --no-auth
```

### CODE REVIEW

**Comment by kurikomoe** [comment]
Ok, I will review the pr this evening (after surviving from my group meeting. XD)

**Comment by kurikomoe** [comment]
Could you provide a `usages.md` in the root folder with the following content?
```markdown
# 各种推理引擎的使用说明

## ollama

### 介绍
原项目地址：https://...
ollama 是 xxxx
优势：
1. xxx
2.

劣势：
1.

### 支持的模型类型
// 这部分写 ollama 支持加载哪些模型，如果是非 SakuraLLM 支持的格式的话，哪些地方能找到第三方维护的转换后的格式。
ollama 私有格式模型，转换方法参见。。

Sakura 相关模型地址：https://

### 前置需求
// 这部分写除了 `requirements.txt` 中 package 依赖以外的依赖。我这边测试自动下载模型是必须安装 ollama 的可执行文件。
1. 需要预装 ollama 程序，用于自动下载 ollama 模型，安装方法参见 xxx
2.

### 使用例子
// 这部分写一下如何启动 ollama server 的。
%```python
python server.py \
    --model_name_or_path onekuma/sakura-13b-lnovel-v0.9b-q2_k \
    --ollama \
    --model_version 0.9 \
    --trust_remote_code \
    --no-auth
%```

### server.py 相关参数及说明
// 这部分写一下 ollama 模型用到了哪些参数，由于早期设计问题，server.py 没有对参数归类，导致混杂在一起了。
%```shell
--model_name_or_path: path or name  [ollama 模型地址]
--model_version: 0.9, xxx  [支持的版本]
--trust_remote_code: 允许不安全代码

--ollama: 启动 ollama 推理引擎
%```
```

**Comment by Isotr0py** [comment]
OK, I will add a `usage.md` later.

File: utils/model.py [hunk 1]
```diff
@@ -72,6 +76,43 @@ def load_model(args: SakuraModelConfig):
     if args.llama_cpp:
         from llama_cpp import Llama
 
+    if args.ollama:
+        import ollama
+        from tqdm import tqdm
+
+        class Ollama:
+            '''Llama-style wrapper for ollama'''
+            def __init__(self, model):
```
[the remark this answers is not in the document]
Comment (Isotr0py) [reply utils/model.py:85]: Maybe we can also implement `install()` method to run install cmd for linux. But it seems that `ollama` in windows and macos only have binary executable installer.

[the remark this answers is not in the document]
Comment (kurikomoe) [reply utils/model.py:85]: Non-explicit installation is not safe (especially when you invoke an external scripts).  Any third-party dependencies should be written inside the README and be installed manually by users.
Besides, the official installation script tends to install `ollama` into `/usr/local/bin`, which requires root permission.

So I think it's better to add a detection here and use the error messages to notify user to read the `usages.md` and install the dependencies themselves.

**Comment by Isotr0py** [comment]
I have added `usage.md`.

I also separate the `ollama.py` to auto invoke `ollama server` command through `multiprocessing`.

**Comment by kurikomoe** [comment]
Sry, my fault. I meant that it's ok for you to only write the ollama part in `usage.md` and I would fill up the other parts myself. Anyway, thanks for your hard work. 

I will review the pr soon.

**Review by kurikomoe — APPROVED** [review]
LGTM

**Pull request merged**